patches/server/0919-Properly-handle-experience-dropping-on-block-break.patch

```diff
@@ -0,0 +1,84 @@
++++ b/src/main/java/net/minecraft/world/level/block/Block.java
+@@ -355,11 +355,16 @@ public class Block extends BlockBehaviour implements ItemLike {
+     // Paper end
+ 
+     public static void dropResources(BlockState state, Level world, BlockPos pos, @Nullable BlockEntity blockEntity, Entity entity, ItemStack stack) {
++    // Paper start
++        dropResources(state, world, pos, blockEntity, entity, stack, true);
++    }
++    public static void dropResources(BlockState state, Level world, BlockPos pos, @Nullable BlockEntity blockEntity, Entity entity, ItemStack stack, boolean dropExp) {
++    // Paper end
+         if (world instanceof ServerLevel) {
+             Block.getDrops(state, (ServerLevel) world, pos, blockEntity, entity, stack).forEach((itemstack1) -> {
+                 Block.popResource(world, pos, itemstack1);
+             });
+-            state.spawnAfterBreak((ServerLevel) world, pos, stack, true);
++            state.spawnAfterBreak((ServerLevel) world, pos, stack, dropExp); // Paper
+         }
+ 
+     }
+@@ -438,9 +443,14 @@ public class Block extends BlockBehaviour implements ItemLike {
+     }
+ 
+     public void playerDestroy(Level world, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack stack) {
++    // Paper start
++        playerDestroy(world, player, pos, state, blockEntity, stack, true);
++    }
++    public void playerDestroy(Level world, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack stack, boolean dropExp) {
++    // Paper end
+         player.awardStat(Stats.BLOCK_MINED.get(this));
+         player.causeFoodExhaustion(0.005F, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.BLOCK_MINED); // CraftBukkit - EntityExhaustionEvent
+-        Block.dropResources(state, world, pos, blockEntity, player, stack);
++        Block.dropResources(state, world, pos, blockEntity, player, stack, dropExp); // Paper
+     }
+ 
+     public void setPlacedBy(Level world, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack itemStack) {}
+diff --git a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
+index 771c6cf992664b65ffbf4ae0192bc7b09f77c2e6..caf924b50b9a3bb0635758873f27a99f66ed6abf 100644
+--- a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
++++ b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
+@@ -1022,6 +1022,7 @@ public abstract class BlockBehaviour {
+ 
+         public void spawnAfterBreak(ServerLevel world, BlockPos pos, ItemStack stack, boolean dropExperience) {
+             this.getBlock().spawnAfterBreak(this.asState(), world, pos, stack, dropExperience);
++            getBlock().popExperience(world, pos, this.getBlock().getExpDrop(asState(), world, pos, stack, true)); // Paper - spawn experience
```

Comment:
doesn't this just totally undo spigot's separation of exp drops and other spawnAfter behaviors? also you aren't checking the boolean, but I think that's irrelevant as this probably shouldn't exist.

Reply:
This is supposed to check on true, the idea was that this would respect it and only not drop xp if it's explicitly needed.